firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java

```diff
@@ -203,9 +225,17 @@ private void initialize(Context context, User user, boolean usePersistence) {
     if (usePersistence) {
       LocalSerializer serializer =
           new LocalSerializer(new RemoteSerializer(databaseInfo.getDatabaseId()));
-      persistence =
+      LruGarbageCollector.Params params = LruGarbageCollector.Params.WithCacheSize(cacheSizeBytes);
+      SQLitePersistence sqlitePersistence =
           new SQLitePersistence(
-              context, databaseInfo.getPersistenceKey(), databaseInfo.getDatabaseId(), serializer);
+              context,
+              databaseInfo.getPersistenceKey(),
+              databaseInfo.getDatabaseId(),
+              serializer,
+              params);
+      lruDelegate = sqlitePersistence.getReferenceDelegate();
+      persistence = sqlitePersistence;
+      scheduleLruGarbageCollection();
```

Comment:
Do we want to check if the cache size is unlimited, so that we can skip the GC run?

Reply:
We can look at this more in the future, but it might actually make sense to split these up, using the same argument that you provided: In FirebaseFirestoreSettings/FirestoreClient we can validate whether the settings for the GC algorithm make sense. In the GC code, we can then always assume that all values make sense and we don't have to special case anything.

That's just my two cents and your mileage may vary. I believe that for now the PR could be simplified if we didn't have to differentiate between -1 and >0 in the LRU GC.

Reply:
I'm not wild about `FirestoreClient` having to know about the constant for whether GC is enabled or not, but I think we can compromise here. What if we gate scheduling on a call to `garbageCollector.isEnabled()` or something like that, as well as add an `assert` to the actual run of the garbage collector, asserting that the minimum threshold is `> 0`? That way, if we ever decide to do something like allow enabling and disabling of garbage collection, the source of truth is still localized to the garbage collector, we'll have asserts if we get something wrong, and we won't fire the timer if we don't need it.

What do you think?

Reply:
Hm, I think that is indeed a compromise. I think I prefer the current solution over asking the GC whether it is enabled and then acting upon this by asking GC to do its job.

I thought of one other approach though - can we passed the queue to the GC algorithm and let it schedule itself if it needs to?

Reply:
I actually initially implemented this, then decided against it because I didn't want to deal with getting timestamps. However, I reworked it when adding the scheduling to `FirestoreClient` so that scheduling is a bit simpler now. It might work well enough to add it back to the garbage collector. I'll try it and we can look at the diff and see if it ends up better.

Reply:
@schmidt-sebastian I started implementing this, but in the process of doing the plumbing to get the `AsyncQueue` to the garbage collector, I decided I really don't like it. A lot of tests that currently run synchronously and have no need of even the notion of our worker thread would have to have an async queue because they happen to use persistence. 

Additionally, I believe the only component that is initialized currently with an async queue is the remote store. I don't think we want to extend that list to include persistence, which is also sort of an end run around sync engine and local store having an async queue.

My next proposal is this instead: what about an object that is solely responsible for the GC scheduling, owned by `FirestoreClient`? My model for doing something like this is `ExponentialBackoff`. `FirestoreClient` would initialize one of these, and then cancel it on shutdown, but would otherwise ignore it.

Reply:
Hey :) I like the idea of following the pattern of `ExponentialBackoff`. Please consider how much boilerplate this might add though, as the current solution might just be a good enough "compromise". If you do want to pursue this and it turns out that reworking the initialization adds a considerable amount of code, then maybe we should revisit this at a later point. 

FWIW, the Web JS SDK does have a dependency on the AsyncQueue from the persistence layer (but that doesn't mean that we have to follow this example).

Reply:
I took a quick stab at it: https://github.com/firebase/firebase-android-sdk/pull/111

I'm not sure it's better, maybe just different? Let me know what you think. If you think it's better, then I'll clean it up and we can merge it, otherwise let's keep what's here.

Reply:
I do like this quite a bit, since it moves the LRU state out of Firestore Client and nicely encapsulates the different timeouts and its scheduling. If you have some time to clean it up (not sure how much you were thinking?) and merge it into here, then I'd be very happy :)

Reply:
OK. I'll merge it here shortly.